### ISSUE TITLE
update controller-runtime

### ISSUE THREAD

**Issue by chuckha** [kubernetes-sigs/cluster-api#1025]
/kind feature

This is a tracking issue for people to pay attention to the controller-runtime dependency update.

**Anything else you would like to add:**
Add related issues below and why they are related:

* https://github.com/kubernetes-sigs/cluster-api/issues/296 -- crd web hook validations are coming with the next version of controller runtime.

**Comment by alexeldeib** [comment]
Are we saying migrating to v2 is dependent on v2 kubebuilder/controller runtime supporting webhooks fully? Or that we don’t have it today so we can punt to post-upgrade

**Comment by detiber** [comment]
I don't necessarily think that it is a blocker, but rather a nice to have.

**Comment by alexeldeib** [comment]
I started looking at the cost of this w.r.t code-generation as well (they are fairly tightly linked, and have dependency interconnection problems). The CAPI side is a decent amount of code but mostly constrained to clusterclient and friends. I haven't looked at it from the provider side yet. 

If the provider side isn't too crazy and most of the work is in clusterclient it might be possible to pull this off fairly smoothly even with ongoing data model changes.

**Comment by tamalsaha** [comment]
Folks, @tahsinrahman and @faem from our team would like to help with this process. We are also blocked on this, as we want to use the client-go 1.14 version for the cloud providers. If you can show us how you like to see this done for one the providers, we can send prs for other ones.

Are these projects planning to switch to the new v2 project structure and go.mod at the same time?

**Comment by detiber** [comment]
I don't have any objections to switching to the v2 project structure and go.mod.

**Comment by alexeldeib** [comment]
I started work to move ClusterClient to use the controller-runtime clients and simultaneously bump to controller-runtime v0.2.0-beta.2. It's very WIP at this stage, I got the builds and type generation generally working. I attempted to directly rip out controller-tools and anything that changed in v1 -> v2 while still supporting `dep` before doing any additional sort of clean up work around client instantiation, etc. 

That said, I'm happy to see someone else go ahead with this instead, as I won't have time to finish it for at least ~a week.

**Comment by vincepri** [comment]
Do Kubernetes and controller-runtime v2 use go.mod?

**Comment by tamalsaha** [comment]
@vincepri , yes.

**Comment by tamalsaha** [comment]
@vincepri , Kubernetes 1.14 does not use go.mod. Kubernetes 1.15 uses go.mod.

kubebuilder's 2.0.0-alpha.X release uses go.mod and k8s 1.14. 

I wonder if we could update kubebuilder and controller-runtime to k8s 1.15 and then update CAPI stuff to 1.15 at the same time. This might be a little ambitious.

**Comment by vincepri** [comment]
That'd be all great, I know there has been a lot of work making sure go.mod works fine with upstream k/k. I only want to make sure that we don't duplicate the burden of supporting go.mod too early for CAPI and all the downstream providers.

**Comment by tamalsaha** [comment]
According to https://github.com/golang/go/wiki/Go-Release-Cycle, GO 1.13 is supposed to be released on Aug 1. My understanding is that it will have go mod on by default. This is probably a good time to switch to go.mod from that point.

**Comment by vincepri** [comment]
Agreed!

**Comment by tamalsaha** [comment]
It seems controller-runtime can be upgraded to use k8s 1.1.5 https://github.com/kubernetes-sigs/controller-runtime/pull/495 . All tests are passing.

**Comment by vincepri** [comment]
@tamalsaha @alexeldeib It seems that to introduce v1alpha2 types (#1035) we'd need to update controller-runtime and controller-tools. Would it be ok for me to open a PR later today to tackle these updates as they're blocking for all the other work items related to v1alpha2?

**Comment by alexeldeib** [comment]
music to my ears. I am looking less at the k8s dependency side and more looking at cutting out the code-gen typed clients in favor of controller-runtime.

**Comment by alexeldeib** [comment]
if anyone is interested, I'm working on that here: https://github.com/kubernetes-sigs/cluster-api/compare/master...alexeldeib:ace/ctrl?expand=1 (very messy WIP branch, but should be clear what I am trying to achieve from clusterclient.go)

**Comment by tamalsaha** [comment]
I have opened a pr against controller-tools to update it to 1.15.0 https://github.com/kubernetes-sigs/controller-tools/pull/231 . One new thing is that instead of one `Webhook` struct, now there are separate `MutatingWebhook` and `ValidatingWebhook`.

**Comment by vincepri** [comment]
Should I push the changes as part of the v1alpha2 types or a different PR? Given that both runtime and tools are beta we probably shouldn’t backport to v1a1 anyways

**Comment by detiber** [comment]
I'm partial to separating them to make reviewing the changes easier

**Comment by vincepri** [comment]
Not a big deal either way, the update currently blocks v1alpha2 so take priority

**Comment by vincepri** [comment]
PR is up, ptal
